<!DOCTYPE html><html lang="zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0,viewport-fit=cover"><title>遇到的问题 | Hexo</title><meta name="author" content="CJ"><meta name="copyright" content="CJ"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="问题一 现象：服务端192.168.101.174使用kinit admin进行认证，可以成功；但是三个客户端192.168.101.179&#x2F;180&#x2F;181都会报错如下： 1kinit: Cannot contact any KDC for realm &amp;#x27;IOTMARS.COM&amp;#x27; while getting initial credentials 原因：KD">
<meta property="og:type" content="article">
<meta property="og:title" content="遇到的问题">
<meta property="og:url" content="http://example.com/2023/05/06/%E6%9D%83%E9%99%90%E8%AE%A4%E8%AF%81/%E9%81%87%E5%88%B0%E7%9A%84%E9%97%AE%E9%A2%98/index.html">
<meta property="og:site_name" content="Hexo">
<meta property="og:description" content="问题一 现象：服务端192.168.101.174使用kinit admin进行认证，可以成功；但是三个客户端192.168.101.179&#x2F;180&#x2F;181都会报错如下： 1kinit: Cannot contact any KDC for realm &amp;#x27;IOTMARS.COM&amp;#x27; while getting initial credentials 原因：KD">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://i.loli.net/2021/02/24/5O1day2nriDzjSu.png">
<meta property="article:published_time" content="2023-05-06T05:31:21.075Z">
<meta property="article:modified_time" content="2023-05-06T05:31:21.075Z">
<meta property="article:author" content="CJ">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://i.loli.net/2021/02/24/5O1day2nriDzjSu.png"><link rel="shortcut icon" href="/img/favicon.png"><link rel="canonical" href="http://example.com/2023/05/06/%E6%9D%83%E9%99%90%E8%AE%A4%E8%AF%81/%E9%81%87%E5%88%B0%E7%9A%84%E9%97%AE%E9%A2%98/index.html"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox/fancybox.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false},
  copy: {
    success: '复制成功',
    error: '复制错误',
    noSupport: '浏览器不支持'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  dateSuffix: {
    just: '刚刚',
    min: '分钟前',
    hour: '小时前',
    day: '天前',
    month: '个月前'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isAnchor: false,
  percent: {
    toc: true,
    rightside: false,
  }
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: '遇到的问题',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2023-05-06 13:31:21'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
    win.getCSS = (url,id = false) => new Promise((resolve, reject) => {
      const link = document.createElement('link')
      link.rel = 'stylesheet'
      link.href = url
      if (id) link.id = id
      link.onerror = reject
      link.onload = link.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        link.onload = link.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(link)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
    const detectApple = () => {
      if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
    })(window)</script><meta name="generator" content="Hexo 6.3.0"></head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src="https://i.loli.net/2021/02/24/5O1day2nriDzjSu.png" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="sidebar-site-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">419</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">0</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">38</div></a></div><hr/></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header"><nav id="nav"><span id="blog-info"><a href="/" title="Hexo"><span class="site-name">Hexo</span></a></span><div id="menus"><div id="toggle-menu"><a class="site-page" href="javascript:void(0);"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">遇到的问题</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">发表于</span><time class="post-meta-date-created" datetime="2023-05-06T05:31:21.075Z" title="发表于 2023-05-06 13:31:21">2023-05-06</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">更新于</span><time class="post-meta-date-updated" datetime="2023-05-06T05:31:21.075Z" title="更新于 2023-05-06 13:31:21">2023-05-06</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E6%9D%83%E9%99%90%E8%AE%A4%E8%AF%81/">权限认证</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title="遇到的问题"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">阅读量:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><h3 id="问题一"><a href="#问题一" class="headerlink" title="问题一"></a>问题一</h3><ul>
<li>现象：服务端192.168.101.174使用kinit admin进行认证，可以成功；但是三个客户端192.168.101.179&#x2F;180&#x2F;181都会报错如下： <figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">kinit: Cannot contact any KDC for realm &#x27;IOTMARS.COM&#x27; while getting initial credentials</span><br></pre></td></tr></table></figure></li>
<li>原因：KDC所在的节点开启了防火墙，导致88端口无法被访问</li>
<li>解决：最简单的就是关闭防火墙。</li>
</ul>
<br>
### 问题二
- 现象：Hadoop整合Kerberos，启动hdfs。访问Namenode Web，发现没有任何Datanode注册到Namenode。随后查看Namenode日志报错如下。
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">2021-09-29 10:57:48,414 WARN org.apache.hadoop.ipc.Client: Couldn&#x27;t setup connection for dn/bigdata1@IOTMARS.COM to bigdata1/192.168.101.179:9820</span><br><span class="line">org.apache.hadoop.ipc.RemoteException(javax.security.sasl.SaslException): GSS initiate failed</span><br></pre></td></tr></table></figure>
- 原因：Datanode无法连接到Namenode的端口。我用的是jdk1.8.0_144，查看相关资料发现是因为jce的问题。jce版本问题，需要网上下载并进行替换。
- 解决：
1、jce下载地址：https://www.oracle.com/technetwork/java/javase/downloads/jce8-download-2133166.html
2、将解压得到的local_policy.jar和US_export_policy.jar拷贝到$JAVA_HOME/jre/lib/security目录下面
3、重启HDFS服务
启动NodeManager时同样会遇到注册失败的问题，解决方法相同。

<br>
### 问题三
- 现象：启动namenode正常，启动datanode时报错无法注册到namenode
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">org.apache.hadoop.hdfs.server.datanode.DataNode: Problem connecting to server:cos-bigdata-test-hadoop-01/192.168.101.184:9820</span><br></pre></td></tr></table></figure>
- 原因：执行命令`netstat -nap | grep 9820`，发现本机监控的地址为**回环地址127.0.0.1**，而不是192.168.101.184，导致其他节点无法进行通讯。
- 解决：修改host文件，在第一行添加  `[本机ip]  [hostname]`
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">192.168.101.186   cos-bigdata-test-hadoop-03</span><br><span class="line">127.0.0.1   localhost localhost.localdomain localhost4 localhost4.localdomain4</span><br><span class="line">::1         localhost localhost.localdomain localhost6 localhost6.localdomain6</span><br><span class="line">192.168.101.184 cos-bigdata-test-hadoop-01 bigdata1</span><br><span class="line">192.168.101.185 cos-bigdata-test-hadoop-02 bigdata2</span><br><span class="line">192.168.101.186 cos-bigdata-test-hadoop-03 bigdata3</span><br></pre></td></tr></table></figure>
如果进程启动正常，但是一直无法注册到namenode或resourcemanager，查看端口发现状态是time_wait，但是查看日志没有报错。这种情况有可能是hosts文件配置出现了问题。


<br>
### 问题四
- 现象：hadoop配置kerberos后，无法启动namenode，datanode等进程。
- 原因：查看log文件后发现是配置的**_HOST**通配符解析有误。
- 解决：_Host文件不是读取的本机的hostname，不同的进程有不同的规则，如下
1. 启动Datanode时，会调用KerberosUtil.java中的getLocalHostName()方法
   <figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">/* Return fqdn of the current host */</span><br><span class="line">static String getLocalHostName() throws UnknownHostException &#123;</span><br><span class="line">  return InetAddress.getLocalHost().getCanonicalHostName();</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
   可以发现使用的是java.net包中的getCanonicalHostName()方法获取全限定主机名。需要在/etc/hosts中的第一行添加  
   <figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">[本机ip] [hostname]</span><br></pre></td></tr></table></figure>
   各个节点启动时会读取这个hostname并赋值给_HOST。

<ol start="2">
<li><p>启动SecondaryNameNode时，会读取配置中的值</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">&lt;property&gt;</span><br><span class="line">    &lt;name&gt;dfs.namenode.secondary.http-address&lt;/name&gt;</span><br><span class="line">    &lt;value&gt;cos-bigdata-test-hadoop-03:9868&lt;/value&gt;</span><br><span class="line">&lt;/property&gt;</span><br></pre></td></tr></table></figure>
<p>启动SecondaryNameNode就会将_HOST设置为cos-bigdata-test-hadoop-03。</p>
</li>
<li><p>启动NameNode时，会读取配置中的值</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">&lt;property&gt;</span><br><span class="line">    &lt;name&gt;fs.defaultFS&lt;/name&gt;</span><br><span class="line">    &lt;value&gt;hdfs://cos-bigdata-test-hadoop-01:9820&lt;/value&gt;</span><br><span class="line">&lt;/property&gt;</span><br></pre></td></tr></table></figure>
<p>启动NameNode就会将_HOST设置为cos-bigdata-test-hadoop-01。</p>
</li>
<li><p>启动ResouceManager时，会读取配置中的值</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">&lt;property&gt;</span><br><span class="line">    &lt;name&gt;yarn.resourcemanager.hostname&lt;/name&gt;</span><br><span class="line">    &lt;value&gt;cos-bigdata-test-hadoop-02&lt;/value&gt;</span><br><span class="line">&lt;/property&gt;</span><br></pre></td></tr></table></figure>
<p>启动ResouceManager就会将_HOST设置为cos-bigdata-test-hadoop-02。</p>
</li>
<li><p>启动jobhistoryserver时，会读取配置中的值</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">&lt;property&gt;</span><br><span class="line">    &lt;name&gt;mapreduce.jobhistory.address&lt;/name&gt;</span><br><span class="line">    &lt;value&gt;cos-bigdata-test-hadoop-03:10020&lt;/value&gt;</span><br><span class="line">&lt;/property&gt;</span><br><span class="line">&lt;property&gt;</span><br><span class="line">    &lt;name&gt;mapreduce.jobhistory.webapp.address&lt;/name&gt;</span><br><span class="line">    &lt;value&gt;cos-bigdata-test-hadoop-03:19888&lt;/value&gt;</span><br><span class="line">&lt;/property&gt;</span><br></pre></td></tr></table></figure>
<p>启动ResouceManager就会将_HOST设置为cos-bigdata-test-hadoop-03。</p>
</li>
<li><p>启动hiveserver2时</p>
</li>
</ol>
<br>
### 问题五
- 现象：本机使用火狐浏览器访问需要Kerberos认证的Namenode Web，出现错误
   <figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">2021-10-15 09:57:18,297 WARN org.apache.hadoop.security.authentication.server.AuthenticationFilter: Authentication exceon: No valid credentials provided (Mechanism level: Failed to find any Kerberos credentails)</span><br></pre></td></tr></table></figure>
- 原因：通过页面查询hadoop目录时，会先验证携带的凭证是否有效，然后向客户端的凭证中写入新的凭证名称，新的凭证名称为 `HTTP/节点名@REALM`，该主机名需要与在hdfs-site中配置的HTTP凭证相同。如果出现节点名不一致的情况，就无法获取该凭证，也就不能访问Web。
![image.png](遇到的问题.assets\37794c07343944f494282ccf9eb05322.png)

<ul>
<li>解决：修改hosts文件，将访问的页面的ip对应的节点名称改为与配置的主体相同的名称。<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">192.168.101.184	cos-bigdata-test-hadoop-01 bigdata1</span><br><span class="line">192.168.101.185	cos-bigdata-test-hadoop-02 bigdata2</span><br><span class="line">192.168.101.186	cos-bigdata-test-hadoop-03 bigdata3</span><br></pre></td></tr></table></figure>
然后清除火狐浏览器的Cookie和网站数据后，重新访问Web。</li>
</ul>
<br>
### 问题六
- 现象：使用命令`hadoop jar /opt/module/hadoop-3.1.3/share/hadoop/mapreduce/hadoop-mapreduce-examples-3.1.3.jar pi 1 1`执行任务时出错，报错如下
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line">[2021-10-14 17:16:43.462]Application application_1634202938998_0001 initialization failed (exitCode=255) with output: main : command provided 0</span><br><span class="line">main : run as user is hxr</span><br><span class="line">main : requested yarn user is hxr</span><br><span class="line">User hxr not found</span><br><span class="line"></span><br><span class="line">2021-10-14 17:16:55,615 INFO mapreduce.Job:  map 100% reduce 100%</span><br><span class="line">2021-10-14 17:16:56,633 INFO mapreduce.Job: Job job_1634202938998_0001 failed with state FAILED due to: Task failed task_1634202938998_0001_m_000000</span><br><span class="line">Job failed as tasks failed. failedMaps:1 failedReduces:0 killedMaps:0 killedReduces: 0</span><br><span class="line"></span><br><span class="line">2021-10-14 17:16:56,724 INFO mapreduce.Job: Counters: 13</span><br><span class="line">	Job Counters </span><br><span class="line">		Failed map tasks=4</span><br><span class="line">		Killed reduce tasks=1</span><br><span class="line">		Launched map tasks=4</span><br><span class="line">		Other local map tasks=3</span><br><span class="line">		Data-local map tasks=1</span><br><span class="line">		Total time spent by all maps in occupied slots (ms)=8940</span><br><span class="line">		Total time spent by all reduces in occupied slots (ms)=0</span><br><span class="line">		Total time spent by all map tasks (ms)=4470</span><br><span class="line">		Total vcore-milliseconds taken by all map tasks=4470</span><br><span class="line">		Total megabyte-milliseconds taken by all map tasks=4577280</span><br><span class="line">	Map-Reduce Framework</span><br><span class="line">		CPU time spent (ms)=0</span><br><span class="line">		Physical memory (bytes) snapshot=0</span><br><span class="line">		Virtual memory (bytes) snapshot=0</span><br><span class="line">Job job_1634202938998_0001 failed!</span><br></pre></td></tr></table></figure>
- 原因：原因可能如下
1. kerberos认证的用户会通过core-site.xml中匹配的规则映射到主机上的用户，主机上的用户必须存在。我这里出现问题的原因就是只有提交的节点上有该用户，但是分布式执行的其他节点上没有该用户。
2. 该用户的id号小于1000或在container-executor.cfg中ban了该用户。

<ul>
<li>解决：</li>
</ul>
<ol>
<li>在其他节点添加该用户；</li>
<li>修改container-executor.cfg配置。</li>
</ol>
<br>
### 问题七
- 现象：启动SecondaryNameNode后，查看日志发现报错如下
   <figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">2021-10-15 10:52:08,802 ERROR org.apache.hadoop.hdfs.server.namenode.SecondaryNameNode: Exception in doCheckpoint</span><br><span class="line">javax.net.ssl.SSLHandshakeException: Error while authenticating with endpoint: https://cos-bigdata-test-hadoop-01:9871/imagetransfer?getimage=1&amp;txid=4419&amp;storageInfo=-64:2014770126:1634089984720:CID-b3e0e3d0-2f3d-49f2-a41c-5725d56b89b5&amp;bootstrapstandby=false</span><br><span class="line">...</span><br><span class="line">Caused by: javax.net.ssl.SSLHandshakeException: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target</span><br></pre></td></tr></table></figure>
- 原因：Https配置时，使用的是keytool生成的证书，导致2nn请求https://cos-bigdata-test-hadoop-01:9871时，证书不可信。
   <figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">[root@cos-bigdata-test-hadoop-03 namesecondary]# curl https://bigdata1:9871</span><br><span class="line">curl: (60) Issuer certificate is invalid.</span><br><span class="line">More details here: http://curl.haxx.se/docs/sslcerts.html</span><br><span class="line"></span><br><span class="line">curl performs SSL certificate verification by default, using a &quot;bundle&quot;</span><br><span class="line">  of Certificate Authority (CA) public keys (CA certs). If the default</span><br><span class="line">  bundle file isn&#x27;t adequate, you can specify an alternate file</span><br><span class="line">  using the --cacert option.</span><br><span class="line"> If this HTTPS server uses a certificate signed by a CA represented in</span><br><span class="line">  the bundle, the certificate verification probably failed due to a</span><br><span class="line">  problem with the certificate (it might be expired, or the name might</span><br><span class="line">  not match the domain name in the URL).</span><br><span class="line"> If you&#x27;d like to turn off curl&#x27;s verification of the certificate, use</span><br><span class="line">  the -k (or --insecure) option.</span><br></pre></td></tr></table></figure>
- 解决：将自签证书导入JDK的信任域中
   将证书添加到jdk的信任库中，jdk的信任库文件是cacerts。
   <figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">keytool -import -v -trustcacerts -alias cos-bigdata-test-hadoop-01 -file /etc/security/keytab/cos-bigdata-test-hadoop-01.crt -keystore $JAVA_HOME/jre/lib/security/cacerts -storepass changeit</span><br></pre></td></tr></table></figure>

<br>
### 问题八
- 现象：配置ranger hive插件后，hiveserver2拉取ranger admin中的配置一直失败，报错如下
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">2021-10-16T14:49:15,936  WARN [Thread-8] client.RangerAdminRESTClient: Error getting policies. secureMode=true, user=hive/cos-bigdata-test-hadoop-01@IOTMARS.COM (auth:KERBEROS), response=&#123;&quot;httpStatusCode&quot;:401,&quot;statusCode&quot;:0&#125;, serviceName=hivetest</span><br></pre></td></tr></table></figure>
- 原因：在hive插件的配置文件中，policymgr_external_url属性设置为ip，而不是主机名。
- 解决：将policymgr_external_url设置为主机名 `POLICY_MGR_URL=http://cos-bigdata-test-hadoop-01:6080`

<br>
### 问题九
- 现象：配置ranger hive和hdfs插件后，都报错如下：
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line">2021-10-18 09:54:07,514 ERROR org.apache.ranger.plugin.util.PolicyRefresher: PolicyRefresher(serviceName=hdfstest): failed to refresh policies. Will continue to use last known version of policies (-1)</span><br><span class="line">com.sun.jersey.api.client.ClientHandlerException: java.lang.RuntimeException: java.lang.NullPointerException</span><br><span class="line">	at com.sun.jersey.client.urlconnection.URLConnectionClientHandler.handle(URLConnectionClientHandler.java:155)</span><br><span class="line">	at com.sun.jersey.api.client.Client.handle(Client.java:652)</span><br><span class="line">	at com.sun.jersey.api.client.WebResource.handle(WebResource.java:682)</span><br><span class="line">	at com.sun.jersey.api.client.WebResource.access$200(WebResource.java:74)</span><br><span class="line">	at com.sun.jersey.api.client.WebResource$Builder.get(WebResource.java:509)</span><br><span class="line">	at org.apache.ranger.admin.client.RangerAdminRESTClient$3.run(RangerAdminRESTClient.java:134)</span><br><span class="line">	at org.apache.ranger.admin.client.RangerAdminRESTClient$3.run(RangerAdminRESTClient.java:126)</span><br><span class="line">	at java.security.AccessController.doPrivileged(Native Method)</span><br><span class="line">	at javax.security.auth.Subject.doAs(Subject.java:360)</span><br><span class="line">	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1709)</span><br><span class="line">	at org.apache.ranger.admin.client.RangerAdminRESTClient.getServicePoliciesIfUpdated(RangerAdminRESTClient.java:137)</span><br><span class="line">	at org.apache.ranger.plugin.util.PolicyRefresher.loadPolicyfromPolicyAdmin(PolicyRefresher.java:251)</span><br><span class="line">	at org.apache.ranger.plugin.util.PolicyRefresher.loadPolicy(PolicyRefresher.java:191)</span><br><span class="line">	at org.apache.ranger.plugin.util.PolicyRefresher.run(PolicyRefresher.java:161)</span><br><span class="line">Caused by: java.lang.RuntimeException: java.lang.NullPointerException</span><br><span class="line">	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1488)</span><br><span class="line">	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1474)</span><br><span class="line">	at sun.net.www.protocol.http.HttpURLConnection.getHeaderField(HttpURLConnection.java:3018)</span><br><span class="line">	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:489)</span><br><span class="line">	at com.sun.jersey.client.urlconnection.URLConnectionClientHandler._invoke(URLConnectionClientHandler.java:253)</span><br><span class="line">	at com.sun.jersey.client.urlconnection.URLConnectionClientHandler.handle(URLConnectionClientHandler.java:153)</span><br><span class="line">	... 13 more</span><br><span class="line">Caused by: java.lang.NullPointerException</span><br><span class="line">	at java.util.Base64$Encoder.encode(Base64.java:261)</span><br><span class="line">	at java.util.Base64$Encoder.encodeToString(Base64.java:315)</span><br><span class="line">	at sun.net.www.protocol.http.NegotiateAuthentication.setHeaders(NegotiateAuthentication.java:208)</span><br><span class="line">	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1731)</span><br><span class="line">	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1474)</span><br><span class="line">	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480)</span><br><span class="line">	... 15 more</span><br></pre></td></tr></table></figure>
- 原因：
- 解决：


<br>
### 问题十
- 现象：Hiveserver2配置Kerberos后，启动报错
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line">2021-09-30T09:50:25,230 ERROR [main] server.HiveServer2: Error starting priviledge synchonizer: </span><br><span class="line">java.lang.NullPointerException: null</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.startPrivilegeSynchonizer(HiveServer2.java:985) ~[hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.start(HiveServer2.java:726) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.startHiveServer2(HiveServer2.java:1037) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.access$1600(HiveServer2.java:140) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2$StartOptionExecutor.execute(HiveServer2.java:1305) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.main(HiveServer2.java:1149) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_144]</span><br><span class="line">	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_144]</span><br><span class="line">	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_144]</span><br><span class="line">	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_144]</span><br><span class="line">	at org.apache.hadoop.util.RunJar.run(RunJar.java:318) [hadoop-common-3.1.3.jar:?]</span><br><span class="line">	at org.apache.hadoop.util.RunJar.main(RunJar.java:232) [hadoop-common-3.1.3.jar:?]</span><br><span class="line"></span><br><span class="line">2021-09-30T09:51:25,453  WARN [main] server.HiveServer2: Error starting HiveServer2 on attempt 3, will retry in 60000ms</span><br><span class="line">org.apache.hive.service.ServiceException: java.lang.NullPointerException</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.start(HiveServer2.java:729) ~[hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.startHiveServer2(HiveServer2.java:1037) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.access$1600(HiveServer2.java:140) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2$StartOptionExecutor.execute(HiveServer2.java:1305) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.main(HiveServer2.java:1149) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_144]</span><br><span class="line">	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_144]</span><br><span class="line">	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_144]</span><br><span class="line">	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_144]</span><br><span class="line">	at org.apache.hadoop.util.RunJar.run(RunJar.java:318) [hadoop-common-3.1.3.jar:?]</span><br><span class="line">	at org.apache.hadoop.util.RunJar.main(RunJar.java:232) [hadoop-common-3.1.3.jar:?]</span><br><span class="line">Caused by: java.lang.NullPointerException</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.startPrivilegeSynchonizer(HiveServer2.java:985) ~[hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.start(HiveServer2.java:726) ~[hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	... 10 more</span><br></pre></td></tr></table></figure>
- 原因：
- 解决：


<br>
### 问题十一
- 问题：执行`./enable-hive-plugin.sh`启动ranger hive插件时，报错如下
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line">Unable to store password in non-plain text format. Error: [SLF4J: Failed to load class &quot;org.slf4j.impl.StaticLoggerBinder&quot;.</span><br><span class="line">SLF4J: Defaulting to no-operation (NOP) logger implementation</span><br><span class="line">SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.</span><br><span class="line">Exception in thread &quot;main&quot; java.lang.NoClassDefFoundError: org/apache/htrace/core/Tracer$Builder</span><br><span class="line">	at org.apache.hadoop.fs.FsTracer.get(FsTracer.java:42)</span><br><span class="line">	at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:3407)</span><br><span class="line">	at org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:158)</span><br><span class="line">	at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:3474)</span><br><span class="line">	at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:3442)</span><br><span class="line">	at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:524)</span><br><span class="line">	at org.apache.hadoop.fs.Path.getFileSystem(Path.java:365)</span><br><span class="line">	at org.apache.hadoop.security.alias.JavaKeyStoreProvider.initFileSystem(JavaKeyStoreProvider.java:89)</span><br><span class="line">	at org.apache.hadoop.security.alias.AbstractJavaKeyStoreProvider.&lt;init&gt;(AbstractJavaKeyStoreProvider.java:85)</span><br><span class="line">	at org.apache.hadoop.security.alias.JavaKeyStoreProvider.&lt;init&gt;(JavaKeyStoreProvider.java:49)</span><br><span class="line">	at org.apache.hadoop.security.alias.JavaKeyStoreProvider.&lt;init&gt;(JavaKeyStoreProvider.java:41)</span><br><span class="line">	at org.apache.hadoop.security.alias.JavaKeyStoreProvider$Factory.createProvider(JavaKeyStoreProvider.java:100)</span><br><span class="line">	at org.apache.hadoop.security.alias.CredentialProviderFactory.getProviders(CredentialProviderFactory.java:73)</span><br><span class="line">	at org.apache.ranger.credentialapi.CredentialReader.getDecryptedString(CredentialReader.java:74)</span><br><span class="line">	at org.apache.ranger.credentialapi.buildks.createCredential(buildks.java:87)</span><br><span class="line">	at org.apache.ranger.credentialapi.buildks.main(buildks.java:41)</span><br><span class="line">Caused by: java.lang.ClassNotFoundException: org.apache.htrace.core.Tracer$Builder</span><br><span class="line">	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)</span><br><span class="line">	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)</span><br><span class="line">	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)</span><br><span class="line">	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)</span><br><span class="line">	... 16 more]</span><br><span class="line">Exiting plugin installation</span><br></pre></td></tr></table></figure>
- 原因：找不到htrace-core4.jar包中的类，原因是${HADOOP_HOME}/share/hadoop/common/lib中的htrace-core4-4.1.0-incubating.jar版本过低。
- 解决：下载htrace-core4-4.2.0-incubating.jar替换htrace-core4-4.1.0-incubating.jar包。


<h3 id="问题十二"><a href="#问题十二" class="headerlink" title="问题十二"></a>问题十二</h3><ul>
<li>问题：Hadoop集群配置Kerberos并配置Ranger hive插件后，启动hiveserver2获取不到在Ranger admin中配置的策略，报错如下<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">2021-10-20T17:15:02,245  WARN [PolicyRefresher(serviceName=hivetest)-33] client.RangerAdminRESTClient: Error getting Roles. secureMode=true, user=hive/cos-bigdata-test-hadoop-01@IOTMARS.COM (auth:KERBEROS), response=&#123;&quot;httpStatusCode&quot;:401,&quot;statusCode&quot;:401,&quot;msgDesc&quot;:&quot;Authentication Failed&quot;&#125;, serviceName=hivetest</span><br><span class="line">2021-10-20T17:15:02,250  WARN [PolicyRefresher(serviceName=hivetest)-33] client.RangerAdminRESTClient: Error getting policies. secureMode=true, user=hive/cos-bigdata-test-hadoop-01@IOTMARS.COM (auth:KERBEROS), response=&#123;&quot;httpStatusCode&quot;:401,&quot;statusCode&quot;:401,&quot;msgDesc&quot;:&quot;Authentication Failed&quot;&#125;, serviceName=hivetest</span><br></pre></td></tr></table></figure></li>
<li>原因：创建ranger:ranger用户启动的ranger各个进程，但是配置的spnego.service.keytab文件的用户是root:hadoop且权限是660，ranger用户没有操作权限。</li>
<li>解决：将ranger用户另外添加到hadoop组 <code>usermod -a -G hadoop ranger</code></li>
</ul>
<h3 id="问题十三"><a href="#问题十三" class="headerlink" title="问题十三"></a>问题十三</h3><ul>
<li>问题：hive配置了Kerberos和Ranger hive插件后，启动hiveserver2后查看hive.log日志，报错如下：<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br></pre></td><td class="code"><pre><span class="line">2021-10-20T23:58:27,406 ERROR [main] server.HiveServer2: Unable to create HiveServer2 namespace: hiveserver2 on ZooKeeper</span><br><span class="line">org.apache.zookeeper.KeeperException$AuthFailedException: KeeperErrorCode = AuthFailed for /hiveserver2</span><br><span class="line">	at org.apache.zookeeper.KeeperException.create(KeeperException.java:130) ~[zookeeper-3.5.7.jar:3.5.7]</span><br><span class="line">	at org.apache.zookeeper.KeeperException.create(KeeperException.java:54) ~[zookeeper-3.5.7.jar:3.5.7]</span><br><span class="line">	at org.apache.zookeeper.ZooKeeper.create(ZooKeeper.java:1538) ~[zookeeper-3.5.7.jar:3.5.7]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl$11.call(CreateBuilderImpl.java:740) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl$11.call(CreateBuilderImpl.java:723) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.curator.RetryLoop.callWithRetry(RetryLoop.java:109) ~[curator-client-2.13.0.jar:?]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl.pathInForeground(CreateBuilderImpl.java:720) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl.protectedPathInForeground(CreateBuilderImpl.java:484) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl.forPath(CreateBuilderImpl.java:474) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl.forPath(CreateBuilderImpl.java:454) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl.forPath(CreateBuilderImpl.java:44) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.startZookeeperClient(HiveServer2.java:490) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.startPrivilegeSynchonizer(HiveServer2.java:1002) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.start(HiveServer2.java:726) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.startHiveServer2(HiveServer2.java:1037) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.access$1600(HiveServer2.java:140) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2$StartOptionExecutor.execute(HiveServer2.java:1305) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.main(HiveServer2.java:1149) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_144]</span><br><span class="line">	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_144]</span><br><span class="line">	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_144]</span><br><span class="line">	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_144]</span><br><span class="line">	at org.apache.hadoop.util.RunJar.run(RunJar.java:318) [hadoop-common-3.1.3.jar:?]</span><br><span class="line">	at org.apache.hadoop.util.RunJar.main(RunJar.java:232) [hadoop-common-3.1.3.jar:?]</span><br><span class="line">2021-10-20T23:58:27,406 ERROR [main] server.HiveServer2: Error starting priviledge synchonizer: </span><br><span class="line">org.apache.zookeeper.KeeperException$AuthFailedException: KeeperErrorCode = AuthFailed for /hiveserver2</span><br><span class="line">	at org.apache.zookeeper.KeeperException.create(KeeperException.java:130) ~[zookeeper-3.5.7.jar:3.5.7]</span><br><span class="line">	at org.apache.zookeeper.KeeperException.create(KeeperException.java:54) ~[zookeeper-3.5.7.jar:3.5.7]</span><br><span class="line">	at org.apache.zookeeper.ZooKeeper.create(ZooKeeper.java:1538) ~[zookeeper-3.5.7.jar:3.5.7]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl$11.call(CreateBuilderImpl.java:740) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl$11.call(CreateBuilderImpl.java:723) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.curator.RetryLoop.callWithRetry(RetryLoop.java:109) ~[curator-client-2.13.0.jar:?]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl.pathInForeground(CreateBuilderImpl.java:720) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl.protectedPathInForeground(CreateBuilderImpl.java:484) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl.forPath(CreateBuilderImpl.java:474) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl.forPath(CreateBuilderImpl.java:454) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl.forPath(CreateBuilderImpl.java:44) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.startZookeeperClient(HiveServer2.java:490) ~[hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.startPrivilegeSynchonizer(HiveServer2.java:1002) ~[hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.start(HiveServer2.java:726) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.startHiveServer2(HiveServer2.java:1037) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.access$1600(HiveServer2.java:140) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2$StartOptionExecutor.execute(HiveServer2.java:1305) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.main(HiveServer2.java:1149) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_144]</span><br><span class="line">	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_144]</span><br><span class="line">	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_144]</span><br><span class="line">	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_144]</span><br><span class="line">	at org.apache.hadoop.util.RunJar.run(RunJar.java:318) [hadoop-common-3.1.3.jar:?]</span><br><span class="line">	at org.apache.hadoop.util.RunJar.main(RunJar.java:232) [hadoop-common-3.1.3.jar:?]</span><br><span class="line">2021-10-20T23:58:27,407  INFO [main] server.HiveServer2: Shutting down HiveServer2</span><br><span class="line">2021-10-20T23:58:27,407  INFO [main] service.AbstractService: Service:ThriftBinaryCLIService is stopped.</span><br><span class="line">2021-10-20T23:58:27,407  INFO [main] service.AbstractService: Service:OperationManager is stopped.</span><br><span class="line">2021-10-20T23:58:27,407  INFO [main] service.AbstractService: Service:SessionManager is stopped.</span><br><span class="line">2021-10-20T23:58:27,407  INFO [main] thrift.ThriftCLIService: Thrift server has stopped</span><br><span class="line">2021-10-20T23:58:27,408  INFO [main] service.AbstractService: Service:CLIService is stopped.</span><br><span class="line">2021-10-20T23:58:27,409  INFO [main] metastore.HiveMetaStore: 0: Done cleaning up thread local RawStore</span><br><span class="line">2021-10-20T23:58:27,409  INFO [main] HiveMetaStore.audit: ugi=hive/cos-bigdata-test-hadoop-01@IOTMARS.COip=unknown-ip-addr	cmd=Done cleaning up thread local RawStore	</span><br><span class="line">2021-10-20T23:58:27,409  INFO [main] service.AbstractService: Service:HiveServer2 is stopped.</span><br><span class="line">2021-10-20T23:58:27,409  INFO [main] server.HiveServer2: Web UI has stopped</span><br><span class="line">2021-10-20T23:58:27,409  INFO [main] server.HiveServer2: Stopping/Disconnecting tez sessions.</span><br><span class="line">2021-10-20T23:58:27,409  INFO [main] session.SparkSessionManagerImpl: Closing the session manager.</span><br><span class="line">2021-10-20T23:58:27,410  WARN [main] server.HiveServer2: Error starting HiveServer2 on attempt 8, will retry in 60000ms</span><br><span class="line">org.apache.hive.service.ServiceException: org.apache.zookeeper.KeeperException$AuthFailedException: KeeperErrorCode = AuthFailed for /hiveserver2</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.start(HiveServer2.java:729) ~[hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.startHiveServer2(HiveServer2.java:1037) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.access$1600(HiveServer2.java:140) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2$StartOptionExecutor.execute(HiveServer2.java:1305) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.main(HiveServer2.java:1149) [hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_144]</span><br><span class="line">	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_144]</span><br><span class="line">	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_144]</span><br><span class="line">	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_144]</span><br><span class="line">	at org.apache.hadoop.util.RunJar.run(RunJar.java:318) [hadoop-common-3.1.3.jar:?]</span><br><span class="line">	at org.apache.hadoop.util.RunJar.main(RunJar.java:232) [hadoop-common-3.1.3.jar:?]</span><br><span class="line">Caused by: org.apache.zookeeper.KeeperException$AuthFailedException: KeeperErrorCode = AuthFailed for /hiveserver2</span><br><span class="line">	at org.apache.zookeeper.KeeperException.create(KeeperException.java:130) ~[zookeeper-3.5.7.jar:3.5.7]</span><br><span class="line">	at org.apache.zookeeper.KeeperException.create(KeeperException.java:54) ~[zookeeper-3.5.7.jar:3.5.7]</span><br><span class="line">	at org.apache.zookeeper.ZooKeeper.create(ZooKeeper.java:1538) ~[zookeeper-3.5.7.jar:3.5.7]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl$11.call(CreateBuilderImpl.java:740) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl$11.call(CreateBuilderImpl.java:723) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.curator.RetryLoop.callWithRetry(RetryLoop.java:109) ~[curator-client-2.13.0.jar:?]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl.pathInForeground(CreateBuilderImpl.java:720) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl.protectedPathInForeground(CreateBuilderImpl.java:484) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl.forPath(CreateBuilderImpl.java:474) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl.forPath(CreateBuilderImpl.java:454) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.curator.framework.imps.CreateBuilderImpl.forPath(CreateBuilderImpl.java:44) ~[curator-framework-2.12.0.jar:?]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.startZookeeperClient(HiveServer2.java:490) ~[hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.startPrivilegeSynchonizer(HiveServer2.java:1002) ~[hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	at org.apache.hive.service.server.HiveServer2.start(HiveServer2.java:726) ~[hive-service-3.1.2.jar:3.1.2]</span><br><span class="line">	... 10 more</span><br></pre></td></tr></table></figure>
错误会导致hiveserver2无法的10000端口关闭导致无法访问。</li>
<li>原因：ranger-2.2.0配置了hive插件后，会配置hiveserver2连接zookeeper来实现HA高可用，但是连接zookeeper时会有权限问题。</li>
<li>解决：1. 配置不连接zookeeper；2. 解决权限问题。</li>
</ul>
<h3 id="问题十四"><a href="#问题十四" class="headerlink" title="问题十四"></a>问题十四</h3><ul>
<li><p>现象：Kerberos集群中，用户认证为hdfs或chenjie用户，启动hive并执行任务后报错，错误信息如下</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">Failed to execute spark task, with exception &#x27;org.apache.hadoop.hive.ql.metadata.HiveException(Failed to create Spark client for Spark session af52c6c3-0317-4b0f-bac4-63f37b2bfcfe)&#x27;</span><br><span class="line">FAILED: Execution Error, return code 30041 from org.apache.hadoop.hive.ql.exec.spark.SparkTask. Failed to create Spark client for Spark session af52c6c3-0317-4b0f-bac4-63f37b2bfcfe</span><br></pre></td></tr></table></figure>
</li>
<li><p>原因：在配置中，禁止了hdfs用户连接，所以不能使用hdfs用户提交任务。具体看resourcemanager的日志，报错如下</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">Failing this attempt.Diagnostics: [2021-11-22 13:40:52.007]Application application_1637559452985_0002 initialization failed (exitCode=255) with output: main : command provided 0</span><br><span class="line">main : run as user is hdfs</span><br><span class="line">main : requested yarn user is hdfs</span><br><span class="line">Requested user hdfs is banned</span><br></pre></td></tr></table></figure>
<p>而chenjie用户在服务器节点中不存在，这样的用户提交任务同样会导致任务失败。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"> Failing this attempt.Diagnostics: [2021-11-22 13:59:28.807]Application application_1637559452985_0005 initialization failed (exitCode=255) with output: main : command provided 0</span><br><span class="line">main : run as user is chenjie</span><br><span class="line">main : requested yarn user is chenjie</span><br><span class="line">User chenjie not found </span><br></pre></td></tr></table></figure>
<blockquote>
<p>在一个非安全的集群里，Hadoop会把计算用的container分发到集群的所有节点上，并且用已经存在的账号或替代账号例如yarn或mapred用户来启动这些container。但是在一个安全集群里，container必须使用你提交的用户名来分发和启动，所以，这个错误是由于你仅仅添加了KDC里面的chenjie用户这个pricipal，而没有在所有节点建立chenjie账户。所以container的执行器无法启动AM或者Mapper或Reducer。</p>
</blockquote>
</li>
<li><p>解决：第一个错误只需要在container-executor.cfg配置文件的配置banned.users&#x3D;hdfs,yarn,mapred中去掉对应的用户即可。第二个错误只需要在所有节点中<strong>新增该用户到hadoop组</strong>即可。</p>
</li>
</ul>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">文章作者: </span><span class="post-copyright-info"><a href="http://example.com">CJ</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">文章链接: </span><span class="post-copyright-info"><a href="http://example.com/2023/05/06/%E6%9D%83%E9%99%90%E8%AE%A4%E8%AF%81/%E9%81%87%E5%88%B0%E7%9A%84%E9%97%AE%E9%A2%98/">http://example.com/2023/05/06/%E6%9D%83%E9%99%90%E8%AE%A4%E8%AF%81/%E9%81%87%E5%88%B0%E7%9A%84%E9%97%AE%E9%A2%98/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">版权声明: </span><span class="post-copyright-info">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0</a> 许可协议。转载请注明来自 <a href="http://example.com" target="_blank">Hexo</a>！</span></div></div><div class="tag_share"><div class="post-meta__tag-list"></div><div class="post_share"><div class="social-share" data-image="https://i.loli.net/2021/02/24/5O1day2nriDzjSu.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2023/05/06/%E6%9D%83%E9%99%90%E8%AE%A4%E8%AF%81/Security/" title="Security"><div class="cover" style="background: var(--default-bg-color)"></div><div class="pagination-info"><div class="label">上一篇</div><div class="prev_info">Security</div></div></a></div><div class="next-post pull-right"><a href="/2023/05/06/%E7%B3%BB%E7%BB%9F%E5%AE%89%E5%85%A8/Git%E5%9F%BA%E6%9C%AC%E5%91%BD%E4%BB%A4/" title="Git基本命令"><div class="cover" style="background: var(--default-bg-color)"></div><div class="pagination-info"><div class="label">下一篇</div><div class="next_info">Git基本命令</div></div></a></div></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src="https://i.loli.net/2021/02/24/5O1day2nriDzjSu.png" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info__name">CJ</div><div class="author-info__description"></div></div><div class="card-info-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">419</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">0</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">38</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/xxxxxx"><i class="fab fa-github"></i><span>Follow Me</span></a></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>公告</span></div><div class="announcement_content">This is my Blog</div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>目录</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E9%97%AE%E9%A2%98%E4%B8%80"><span class="toc-number">1.</span> <span class="toc-text">问题一</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E9%97%AE%E9%A2%98%E5%8D%81%E4%BA%8C"><span class="toc-number">2.</span> <span class="toc-text">问题十二</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E9%97%AE%E9%A2%98%E5%8D%81%E4%B8%89"><span class="toc-number">3.</span> <span class="toc-text">问题十三</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E9%97%AE%E9%A2%98%E5%8D%81%E5%9B%9B"><span class="toc-number">4.</span> <span class="toc-text">问题十四</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>最新文章</span></div><div class="aside-list"><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2023/05/06/MySQL/%E6%B3%A8%E8%A7%A3@Select%E5%92%8C@Insert/" title="注解@Select和@Insert">注解@Select和@Insert</a><time datetime="2023-05-06T05:48:28.906Z" title="发表于 2023-05-06 13:48:28">2023-05-06</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2023/05/06/%E5%90%8E%E7%AB%AF%E6%A1%86%E6%9E%B6/%E6%B3%A8%E8%A7%A3@EnableAutoConfiguration/" title="注解@EnableAutoConfiguration">注解@EnableAutoConfiguration</a><time datetime="2023-05-06T05:48:06.027Z" title="发表于 2023-05-06 13:48:06">2023-05-06</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2023/05/06/%E5%A4%A7%E6%95%B0%E6%8D%AE%E7%A6%BB%E7%BA%BF/%E5%A4%A7%E6%95%B0%E6%8D%AE%E9%9B%86%E7%BE%A4%E7%9B%91%E6%8E%A7%E6%A1%86%E6%9E%B6/" title="大数据集群监控框架">大数据集群监控框架</a><time datetime="2023-05-06T05:42:56.298Z" title="发表于 2023-05-06 13:42:56">2023-05-06</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2023/05/06/%E9%AB%98%E5%B9%B6%E5%8F%91/HashMap%E5%B9%B6%E5%8F%91%E9%97%AE%E9%A2%98%E5%8F%8AConcurrentHashMap%E5%8E%9F%E7%90%86/" title="HashMap并发问题及ConcurrentHashMap原理">HashMap并发问题及ConcurrentHashMap原理</a><time datetime="2023-05-06T05:31:21.103Z" title="发表于 2023-05-06 13:31:21">2023-05-06</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2023/05/06/%E9%AB%98%E5%B9%B6%E5%8F%91/Stream%E5%8E%9F%E7%90%86/" title="Stream原理">Stream原理</a><time datetime="2023-05-06T05:31:21.103Z" title="发表于 2023-05-06 13:31:21">2023-05-06</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2020 - 2023 By CJ</div><div class="framework-info"><span>框架 </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>主题 </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="阅读模式"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="浅色和深色模式转换"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="单栏和双栏切换"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="设置"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="目录"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="回到顶部"><span class="scroll-percent"></span><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox/fancybox.umd.min.js"></script><div class="js-pjax"></div><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>